### ISSUE TITLE
Add dark theme support to the admin

### ISSUE THREAD

**Issue by Jupi007** [sulu/sulu#6830]
# Problem description

I personally always switch my system to a dark theme at night.
More and more websites are now supporting this feature, and I always implement a dark theme on my projects.
I'd like to build a large part of my next ones using Sulu, and the transition between the public part and the admin is a bit eye hurting at night in my opinion.

# Proposed solution

I would love to see the Sulu admin in dark theme. The `prefers-color-scheme` css feature can be used, which is supported by all modern browser: https://caniuse.com/?search=prefers-color-scheme

```css
@media (prefers-color-scheme: dark) {
...
}
```

I can help by doing a working skeleton base PR that you could fill with your own colors if you want.

**Comment by alexander-schranz** [comment]
We are very strict about the design of the Sulu Admin. It would be first needed to be tackled by our UI/UX experts but as there are more important tasks then a dark theme  
currently there will be not soon a dark theme provided for the Sulu Admin.

I personally also not a big fan of this kind of providing dark or light themes to a software but not tackling something more specific kind of things like a specific kind of color blindness to make things better accessible.

But yeah I'm really not a fan of the dark/light theme movement and really want when this topic is brought up that a additional theme should solve something more then just make things dark or light. As I think accessible themes are more important then have the availability to switch between dark or light mode. But yeah I have a very strict opinion on this kind of topic. But that does not exclude that Sulu will maybe have in future a dark theme, just want to add my point of view on this kind of topic.

If you are currently just search a way how to replace existing colors of Sulu inside your project, have a look at the following issue https://github.com/sulu/sulu/issues/6820 and https://github.com/sulu/sulu-demo/pull/75.

All colors can be found here: https://github.com/sulu/sulu/blob/2.5/src/Sulu/Bundle/AdminBundle/Resources/js/containers/Application/colors.scss. Sadly our naming of colors is not very theming friendly would personally prefered to maybe change this in future in more tailwind way color schema but usage of `primary`, `secondary`, `gray`, `error`, `warning`, ... color names. In the tailwind schema e.g.: `primary-500`, `gray-300`, ...

Still would be interested what you think what would be required for making sulu more theming friendly. Or what you would come up with.

**Comment by Jupi007** [comment]
No problem, I totally understand your position :)

> If you are currently just search a way how to replace existing colors of Sulu inside your project, have a look at the following issue #6820 and [sulu/sulu-demo#75](https://github.com/sulu/sulu-demo/pull/75).

Thanks, but I think this wouldn't work, as I search to have both light/dark themes and let the browser switch to the best one depending on the user choice.
Unless copy and rewrite a large part of your scss, I don't think it would be easily doable.

> Still would be interested what you think what would be required for making sulu more theming friendly. Or what you would come up with.

Well on a first look, I think use root css vars for all that is related to colors would allow a great possibility of extends your admin ui. Because unlike scss vars, css ones allow to mix them with some features like `prefers-color-scheme`.
Also as you said, use semantic var names would be a good plus.

**Comment by alexander-schranz** [comment]
Yeah maybe the first step could be convert existing colors to css variables e.g.:

Something like:

```diff
// colors.scss

-$blueZodiac: #112a46;
+$blueZodiac: var(--color-blue-zodiac)

// global.scss:
+:root {
+    --color-blue-zodiac: #112a46;
+}
```

Think this way it should not be a bc break for custom modules using the scss variables. So we could make things step by step into this direction.

**Comment by Jupi007** [comment]
Exactly.
If you don't use any color transformation in modules (like `opacity`, `lighten`...), it should work.

Then, project extended themes could do something like:

```css
@media (prefers-color-scheme: dark) {
  :root {
    --color-blue-zodiac: #darkerblueshade;
  }
}
```

_Awaiting semantic named vars, because color names will have no more sense in this example._

Btw, if you go in this direction with css vars, it wouldn't be a big deal to add dark theme (or some high contrast mode) support to Sulu admin. You would just need to add dark shade variants of all colors in `global.scss`.

**Comment by alexander-schranz** [comment]
@Jupi007 

> If you don't use any color transformation in modules (like opacity, lighten...), it should work.

We are using postcss and only a limited feature set of scss. Nesting, simple-vars, the only things we need to check is the usage of `rgba` via the [`postcss-hexrgba`](https://github.com/madeleineostoja/postcss-hexrgba) plugin. But yeah that would be needed to be checked how we could achieve this.

I know tailwind uses here `background: rgba(var(--color-primary-500),0.5)`. But the colors are there defined this way `--color-primary-500: 255, 255, 0` and does not include `rgb` or `rgba` so they don't work out of the box e.g. as `color: $blueZodiac` as that would be `color: 255, 255, 0`. Not sure if maybe already a postcss plugin exist to wrap that correctly.

But maybe analyzing where is currently rgba used where it currently really make sense. I know there are some cases where instead of a "gray" rgba($black, 0.8) which would already from theming point of view be false. So maybe a lot of rgba could just be removed and then we just need to fight with the once which stays and maybe make them also css variables if they are not a lot.

**Comment by Jupi007** [comment]
> I know tailwind uses here background: rgba(var(--color-primary-500),0.5). But the colors are there defined this way --color-primary-500: 255, 255, 0 and does not include rgb or rgba so they don't work out of the box e.g. as color: $blueZodiac as that would be color: 255, 255, 0. Not sure if maybe already a postcss plugin exist to wrap that correctly.

Hmm, indeed I see you are using a lot of `rgba` and I don't see how you could easily get rid of them, because most are used for shadows and disabled stuff, without add a lot of complexity to the color palette.

You could do something like this:

```scss
--color-primary-500: 255, 255, 0;
$accent-primary: var(--color-primary-500);
// ...
color: rgb($accent-primary);
color: rgba($accent-primary, .5);
```

But this will break bc, as the color format must follow a precise pattern which isn't compatible with the older one.
And I don't think that a postcss plugin exists to work with `rbg` color without `rgb`.

But imo, doing like in the upper example would be the best solution for theming:
- use css vars everywhere ;
- keep rgba (to hide complexity) ;
- and most important use semantic var names (`$bg-color`, `$fg-color`, `$accent-primary`...), which can be easily overwritten without any side effects.

This mean a bc break, but I think doing an intermediary job isn't a good idea because you will need to adapt a large part of your colors to get rid of `rgba`, to then go back with using it.
